File: 2023_algorithms/golang/quick_select.go
package goimpls

import (
// "fmt"
)

func quickSelectPartition(arr []int, low, high int) int {
	if low >= high {
		return low
	}

	index := low
	pivot := arr[high]

	for i := low; i < high; i++ {
		if arr[i] <= pivot {
			// fmt.Printf("~~~ i=%d, index=%d, arr[i]=%d\n", i, index, arr[i])
			if index != i {
				arr[index], arr[i] = arr[i], arr[index]
			}
			index += 1
		}
	}
	arr[index], arr[high] = arr[high], arr[index]

	return index
}

func quickSelectRange(arr []int, k, low, high int) int {
	if len(arr) == 0 {
		panic("empty array")
	} else if len(arr) == 1 {
		return 0
	}

	if low >= high {
		return high
	}

	pivot := quickSelectPartition(arr, low, high)
	switch {
	case k < pivot:
		return quickSelectRange(arr, k, low, pivot-1)
	case k > pivot:
		return quickSelectRange(arr, k, pivot+1, high)
	default:
		return pivot
	}
}

func QuickSelect(arr []int, k int) int {
	return quickSelectRange(arr, k, 0, len(arr)-1)
}

File: 2024_algorithms/DSAR/p02/src/stats/quickselect.rs
pub fn partition<T: PartialOrd>(array: &mut [T], low: usize, high: usize) -> usize {
    let pivot = high;
    let mut i = low;

    for j in low..high {
        if array[j] <= array[pivot] {
            array.swap(i, j);
            i += 1;
        }
    }

    array.swap(i, high);
    i
}

fn _quickselect<T: PartialOrd>(array: &mut [T], low: usize, high: usize, k: usize) -> &T {
    if low == high {
        return &array[low];
    }

    let pivot_index = partition(array, low, high);

    if k == pivot_index {
        &array[k]
    } else if k < pivot_index {
        _quickselect(array, low, pivot_index - 1, k)
    } else {
        _quickselect(array, pivot_index + 1, high, k)
    }
}

fn quickselect<T: PartialOrd>(array: &mut [T], k: usize) -> Option<&T> {
    if k == 0 || k > array.len() {
        return None;
    }

    Some(_quickselect(array, 0, array.len() - 1, k - 1))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn t_quickselect() {
        let mut data = vec![3, 5, 1, 9, 7, 6, 2, 8, 4];

        // Looking for the 5th smallest element, 1-based index
        assert_eq!(quickselect(&mut data, 4), Some(&4));

        assert_eq!(quickselect(&mut data, 3), Some(&3));

        assert!(quickselect(&mut data, 0).is_none());
        assert_eq!(quickselect(&mut data, 1), Some(&1));

        assert_eq!(quickselect(&mut data, 9), Some(&9));
        assert!(quickselect(&mut data, 10).is_none());
    }
}
